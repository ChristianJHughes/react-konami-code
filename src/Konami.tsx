import React from 'react';

import PropTypes from 'prop-types';

export interface KonamiProps {
  className?: string;
  code?: number[];
  disabled?: boolean;
  resetDelay?: number;
  action?: () => void;
  timeout?: number;
  onTimeout?: () => void;
}

export interface KonamiStates {
  done: boolean;
  input: number[];
}

class Timer {
  t: number;
  timerIntervalID: number | null;
  fn: TimerHandler;

  constructor(fn: () => void, t: number) {
    this.t = t;
    this.fn = fn;

    this.timerIntervalID = setInterval(this.fn, this.t);
  }

  stop() {
    if (this.timerIntervalID) {
      clearInterval(this.timerIntervalID);
      this.timerIntervalID = null;
    }
    return this;
  }

  start() {
    if (!this.timerIntervalID) {
      this.stop();
      this.timerIntervalID = setInterval(this.fn, this.t);
    }
    return this;
  }

  reset(newT: number) {
    this.t = newT;
    return this.stop().start();
  }
}

const propTypes = {
  action: PropTypes.func,
  className: PropTypes.string,
  code: PropTypes.arrayOf(PropTypes.number),
  disabled: PropTypes.bool,
  onTimeout: PropTypes.func,
  resetDelay: PropTypes.number,
  timeout: PropTypes.number,
};

class Konami extends React.Component<KonamiProps, KonamiStates> {
  private timeoutID: ReturnType<typeof setTimeout> | null | undefined;
  private _timer: any;
  static defaultProps: KonamiProps;

  static propTypes: typeof propTypes;

  constructor(props: KonamiProps) {
    super(props);

    this.state = {
      done: false,
      input: [],
    };

    this.timeoutID = null;
    this.onKeyUp = this.onKeyUp.bind(this);
    this.resetInput = this.resetInput.bind(this);
  }

  componentDidMount() {
    const { resetDelay } = this.props;

    document.addEventListener('keyup', this.onKeyUp);
    const delay = Number(resetDelay);
    if (delay !== 0) {
      this._timer = new Timer(() => this.resetInput(), delay);
    }
  }

  componentWillUnmount() {
    const { resetDelay } = this.props;

    if (this.timeoutID) clearTimeout(this.timeoutID);

    if (resetDelay !== 0) {
      this._timer.stop();
    }
    document.removeEventListener('keyup', this.onKeyUp);
  }

  onKeyUp(e: KeyboardEvent) {
    const { done, input } = this.state;
    const {
      action,
      code,
      disabled,
      onTimeout,
      resetDelay,
      timeout,
    } = this.props;

    const delay = Number(resetDelay);

    if (disabled) {
      return;
    }

    if (delay !== 0) {
      this._timer.reset(delay);
    }

    input.push(e.keyCode);

    if (code) input.splice(-code.length - 1, input.length - code.length);

    this.setState({ input }, () => {
      if (this.state.input.join('').includes(code?.join('') ?? '') && !done) {
        if (delay !== 0) {
          this._timer.stop();
        }
        this.setState({ done: true }, () => {
          if (typeof action === 'function') {
            action();
          }
        });

        if (timeout) {
          this.timeoutID = setTimeout(() => {
            this.setState({ done: false });
            if (typeof onTimeout === 'function') {
              onTimeout();
            }
          }, Number(timeout));
        }
      }
    });
  }

  resetInput() {
    this.setState({ input: [] });
  }

  render() {
    const { children, className, disabled } = this.props;
    const { done } = this.state;

    return (
      <div
        className={`konami ${className ?? ''}`}
        style={{ display: !done || disabled ? 'none' : 'block' }}
      >
        {children}
      </div>
    );
  }
}

Konami.defaultProps = {
  className: '',
  code: [38, 38, 40, 40, 37, 39, 37, 39, 66, 65],
  disabled: false,
  resetDelay: 1000,
};

Konami.propTypes = propTypes;

export default Konami;
