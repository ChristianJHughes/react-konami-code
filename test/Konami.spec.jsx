import React from 'react';
import Konami from '../src';

describe('Test cases', () => {
  it('should render default component', () => {
    const wrapper = shallow(<Konami>{'Hey, I\'m an Easter Egg!'}</Konami>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should add a className to the wrapping div', () => {
    const wrapper = shallow(<Konami className="customClass">{'Hey, I\'m an Easter Egg!'}</Konami>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not be done after wrong code input', () => {
    const wrapper = shallow(<Konami>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    const input = [38, 37, 40, 40, 37, 39, 37, 39, 66, 65];
    input.forEach((keyCode) => {
      instance.onKeyUp({ keyCode });
    });

    expect(wrapper.state('done')).toEqual(false);
  });

  it('should be done after code input', () => {
    const cb = jest.fn();
    const wrapper = shallow(<Konami action={cb}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    const input = [38, 38, 40, 40, 37, 39, 37, 39, 66, 65];
    input.forEach((keyCode) => {
      instance.onKeyUp({ keyCode });
    });

    expect(wrapper.state('done')).toEqual(true);
    expect(cb).toHaveBeenCalledTimes(1);
  });

  it('should not be done after code input while disabled', () => {
    const wrapper = shallow(<Konami disabled>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    const input = [38, 38, 40, 40, 37, 39, 37, 39, 66, 65];
    input.forEach((keyCode) => {
      instance.onKeyUp({ keyCode });
    });

    expect(wrapper.state('done')).toEqual(false);
  });

  it('should be done after custom code input', () => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });
    instance.onKeyUp({ keyCode: 55 });

    expect(wrapper.state('done')).toEqual(true);
  });

  it('should not be done after timeout', (done) => {
    const onEnd = jest.fn();
    const wrapper = shallow(
      <Konami code={[55, 55]} onTimeout={onEnd} timeout={1000}>
        {'Hey, I\'m an Easter Egg!'}
      </Konami>,
    );

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      expect(wrapper.state('done')).toEqual(false);
      expect(onEnd).toHaveBeenCalledTimes(1);
      done();
    }, 1000);
  });

  it('should not be done if you are slower than resetDelay', (done) => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      instance.onKeyUp({ keyCode: 55 });
      expect(wrapper.state('done')).toEqual(false);
      done();
    }, 1000);
  });

  it('should be done if you are slow but resetDelay is zero', (done) => {
    const wrapper = shallow(<Konami code={[55, 55]} resetDelay={0}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      instance.onKeyUp({ keyCode: 55 });
      expect(wrapper.state('done')).toEqual(true);
      done();
    }, 1000);
  });

  it('should reset input if user is slow', (done) => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      expect(wrapper.state('input')).toEqual([]);
      done();
    }, 1000);
  });

  it('should unmount component', () => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    const componentWillUnmount = jest.spyOn(instance, 'componentWillUnmount');
    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalled();
  });
});
