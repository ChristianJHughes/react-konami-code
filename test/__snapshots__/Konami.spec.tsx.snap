// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cases should add a className to the wrapping div 1`] = `
<div
  className="konami customClass"
  style={
    Object {
      "display": "none",
    }
  }
>
  Hey, I'm an Easter Egg!
</div>
`;

exports[`Test cases should render default component 1`] = `
<div
  className="konami "
  style={
    Object {
      "display": "none",
    }
  }
>
  Hey, I'm an Easter Egg!
</div>
`;
